[Merged by Bors] - feat(data/sym2): add the universal property, and make `sym2.is_diag ⟦(x, y)⟧ ↔ x = y` true definitionally

eric-wieser opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

eric-wieser committed:
feat(data/sym2): add the universal property, and make `sym2.is_diag ⟦(x, y)⟧ ↔ x = y` true definitionally

pechersky commented on src/data/sym2.lean:
```diff
@@ -200,20 +215,32 @@ def diag (x : α) : sym2 α := ⟦(x, x)⟧
+prod.rec_on z $ λ _ _, is_diag_iff_eq
+
+@[simp]
+lemma diag_is_diag (a : α) : is_diag (diag a) :=
```
```suggestion
lemma is_diag_diag (a : α) : is_diag (diag a) :=
```
And the appropriate edits below

eric-wieser replied:
This name is the same as it was before this PR, but I suppose I could creep the scope to fix the name.

kmill replied:
I get confused about the name to give compositions like this, and I get even more confused when dot notation is involved.  If the conclusion were written as `(diag a).is_diag` (which it could be), would it still be `is_diag_diag`?

I think I'm guilty putting `is_foo` predicates at the ends of lemmas elsewhere, too, which is why I ask.

eric-wieser replied:
I made [a Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/.23naming.20and.20dot.20notation/near/246520191).

eric-wieser replied:
I've also reverted the change - if it's contentious, then renaming it is outside the scope of this PR.

YaelDillies commented on src/data/sym2.lean:
```diff
@@ -94,6 +94,21 @@ begin
+`sym2.from_rel` is a more convenient spelling when working with `Prop`. -/
+def lift {β : Sort*} :
+  {f : α → α → β // ∀ a₁ a₂, f a₁ a₂ = f a₂ a₁} ≃ (sym2 α → β) :=
+{ to_fun := λ f, quotient.lift (λ a : α × α, (f : α → α → β) a.1 a.2) $
```
Could you please make that `to_fun` a separate definition? We needed that generalisation of `sym2.from_rel` some time ago.

eric-wieser replied:
What would be the point of the separate definition? If you use `lift` it unfolds to this anyway.

YaelDillies replied:
Well I guess it's easier to have a definition without mentioning a subtype.
```
def from_fun {β : Sort*} {f : α → α → β} (comm : ∀ ⦃x y⦄, f x y = f y x) :
  sym2 α → β :=
quotient.lift (uncurry f) (by { rintro ⟨x, y⟩ _ ⟨_, _⟩, {refl }, unfold uncurry, rw comm })
```

eric-wieser replied:
That def results in ` {β : Type*}` since `uncurry` isn't defined on sorts:
```lean
#check @from_fun
-- Π {α : Type u_1} {β : Type u_2} {f : α → α → β}, (∀ ⦃x y : α⦄, f x y = f y x) → sym2 α → β
```

YaelDillies replied:
Oh that's annoying... There's no real reason for it, right?

YaelDillies replied:
Wait, what? The original definition of `from_rel` uses `function.uncurry r`. What is this witchcraft?

eric-wieser replied:
Looking again, it looks like there's no point in letting `lift` work on `Sort` anyway, no one every needs a subtype of a proof, and `quotient.induction` works just fine for the cases when Sort would apply.

eric-wieser committed:
rename and golf

eric-wieser commented on src/data/sym2.lean:
```diff
@@ -200,20 +215,32 @@ def diag (x : α) : sym2 α := ⟦(x, x)⟧
+def is_diag : sym2 α → Prop :=
+lift ⟨eq, λ _ _, propext eq_comm⟩
 
 @[simp]
```
```suggestion
```

eric-wieser committed:
lintfix

eric-wieser committed:
use curry, remove Sort*

eric-wieser committed:
Merge remote-tracking branch 'origin/eric-wieser/sym2.lift' into eric-wieser/sym2.lift

kmill commented on src/data/sym2.lean:
```diff
@@ -94,6 +94,19 @@ begin
+`sym2.from_rel` instead. -/
+def lift {β : Type*} : {f : α → α → β // ∀ a₁ a₂, f a₁ a₂ = f a₂ a₁} ≃ (sym2 α → β) :=
+{ to_fun := λ f, quotient.lift (uncurry ↑f) $ by { rintro _ _ ⟨⟩, exacts [rfl, f.prop _ _] },
+  inv_fun := λ F, ⟨λ a₁ a₂, F ⟦(a₁, a₂)⟧, λ a₁ a₂, congr_arg F eq_swap⟩,
```
If we want to maintain the point-free style of `to_fun`, then `inv_fun `could be
```lean
  inv_fun := λ F, ⟨curry (F ∘ quotient.mk), λ a₁ a₂, congr_arg F eq_swap⟩,
```

eric-wieser replied:
Sure, why not. I added a lemma that unfolds it to my version, but it's nice to see the composition in `lift.inv_fun`, since there almost always is one.

kmill reviewed (approved):
`sym2.lift` should have existed long ago -- thanks!

eric-wieser committed:
Change back is_diag_diag

Since there's no consensus, lets leave renaming this for another time.

Also adds a docstring and a lemma

b-mehta reviewed (approved):
Looks good to me! I don't have a strong opinion on the `is_diag_diag` naming choice, so I'll let @eric-wieser decide what to do there.

bors d+
